ejb/WFLY-6892-ejb-access-logging.adoc

```diff
@@ -0,0 +1,138 @@
+[x] Engineering
+
+[x] QE
+
+=== Affected Projects or Components
+
+* ejb3
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* implement the access logging feature similar to WildFly web access logging.
+
+* add access-logging definition to wildfly-ejb3 schema (see next section below)
+
+* configure ejb access logging with CLI
+
+* enable and disable ejb access logging
+
+* configure how to rotate ejb access logs if applicable
+
+** default should be OFF for performance consideration
+
+* configure output location and format
+
+** directory/file
+
+** stdout
+
+* choose from pre-defined logging formats (need to adapt for ejb invocation):
+
+** common (default): %h %l %u %t "%r" %s %b
+
+** combined: %h %l %u %t "%r" %s %b "%{i,Referer}" "%{i,User-Agent}"
+
+* for more customization, define which attributes to include in ejb access logging
+
+** consider adapting the following attributes that are currently supported by web access logging:
+
+    %a - Remote IP address
+    %A - Local IP address
+    %b - Bytes sent, excluding HTTP headers, or '-' if no bytes were sent
+    %B - Bytes sent, excluding HTTP headers
+    %h - Remote host name
+    %H - Request protocol
+    %l - Remote logical username from identd (always returns '-')
+    %m - Request method
+    %p - Local port
+    %q - Query string (excluding the '?' character)
+    %r - First line of the request
+    %s - HTTP status code of the response
+    %t - Date and time, in Common Log Format format
+    %u - Remote user that was authenticated
+    %U - Requested URL path (ejb invocation equivalent)
+    %v - Local server name
+    %D - Time taken to process the request, in millis
+    %T - Time taken to process the request, in seconds
+    %I - current Request thread name (can compare later with stacktraces)
+
+=== New elements in wildfly-ejb3_6_0.xsd schema
+ejb30 subsystem is extended with a new top-level element access-log, defined as follows:
+[source,xml]
+----
+    <xs:complexType name="accessLogType">
+        <xs:attribute name="pattern" use="optional" type="xs:string" default="common"/>
+        <xs:attribute name="worker" use="optional" type="xs:string" default="default"/>
+        <xs:attribute name="directory" use="optional" type="xs:string" default="${jboss.server.log.dir}"/>
+        <xs:attribute name="relative-to" use="optional" type="xs:string" />
+        <xs:attribute name="prefix" use="optional" type="xs:string" default="access_log."/>
+        <xs:attribute name="suffix" use="optional" type="xs:string" default="log"/>
+        <xs:attribute name="rotate" use="optional" type="xs:string" default="true"/>
+        <xs:attribute name="use-server-log" use="optional" type="xs:string" default="false"/>
+        <xs:attribute name="extended" use="optional" type="xs:string" default="false" />
+        <xs:attribute name="predicate" use="optional" type="xs:string" />
+    </xs:complexType>
```

Comment:
I'd like to suggest possibly looking at a different model for this. When working on the JSON and console logging for web access logging this model did not work well for expansion. I'd suggest having separate resources and not making the `use-server-log` mutually exclusive.

Reply:
Can you elaborate on the model that is more friendly to JSON and console logging?  What do you mean by "having separate resources"?  Does it mean an ejb access-log model that is different from web access-log?  Also can you explain more on "not making user-server-log mutually exclusive?"

Reply:
I ended up adding the console logging as a separate resource, `setting=console-access-log`. This was because using new attributes would require too many mutually exclusive attributes. Meaning you would have something like `use-console-logging` which means `use-server-log` can't be `true` as well and it also means that it would turn off the file logging. This doesn't seem very intuitive. I also think users should be allowed to log in as many places as they'd like. For example if they want to use the server logging and log to a file that seems like a valid use case.

I would suggest doing something like, please not I'm guessing on the address,: 
* `/subsystem=ejb3/service=access-logging/json-formatter=my-json-formatter`
* `/subsystem=ejb3/service=access-logging/file=myfile`
* `/subsystem=ejb3/service=access-logging/console=my-console`
* `/subsystem=ejb3/service=access-logging/pattern-formatter=mypattern`
* `/subsystem=ejb3/service=access-logging:write-attribute(name=use-server-log, value=true)`

Note these names can definitely change it's just a concept. The idea though is if you end up with the requirement for an additional output type, say a socket out, you could easily add another resource.

Reply:
Got it.  I'll update it based on the discussion so far.

Reply:
Sounds good. It doesn't have to be specifically those address, but just a way to allow for any output and format to be used on that output.

Note too if you don't use the event logging API some special handling for the console will need to be used as we wrap `System.out` and `System.err` in loggers.

Reply:
I updated the schema segment to model after jboss-as-logging-7_0.xsd, and it ends up quite long addition.  Not sure if there's a better way.  Some schema elements definition differ from jboss-as-logging ones to adapt to access logging.  With this changes, one should be able to specify multiple ejb access logs, each with its own formatter (simple or JSON).

I also defined use-server-log as its own element (instead of an attribute), in order to apply any named or inline formatter.